### ISSUE TITLE
It seems that unexpected result returned in floating-point constraints

### ISSUE THREAD

**Issue by martin2011** [Z3Prover/z3#1120]
Hi, the conflict floating-point constraints seem to be solved as satisfiable. The constraints are as followed:
```
1) !(y>0.0)
2) fabs(x-y)<M, where M is a big floating-point number.
3) x >=0.0
4) sqrt(x-y) != 0.0
5) sqrt(x-y) !=0.0  && 2.0 * sqrt(x-y) == 0.0
```
, and  from the fifth constraint, we known the constraints are `unsat`.  But, we get the 'sat' returned, the constraints are translated as SMTLIB format as following:
```
(declare-fun symbol0 () (_ FloatingPoint 11 53))
(declare-fun symbol1 () (_ FloatingPoint 11 53))
(assert
 (not (fp.isInfinite symbol0)))
(assert
 (not (fp.isNaN symbol0)))
(assert
 (not (fp.isInfinite symbol1)))
(assert
 (not (fp.isNaN symbol1)))
(assert
 (let ((?x14 (fp.sub roundNearestTiesToEven symbol0 symbol1)))
 (let ((?x28 (fp.sqrt roundNearestTiesToEven ?x14)))
 (let (($x9 (and (distinct ?x28 ((_ to_fp 11 53) RTZ +zero)) true)))
 (and $x9 (fp.eq (fp.mul roundNearestTiesToEven ((_ to_fp 11 53) RTZ 1.0 1) ?x28) ((_ to_fp 11 53) RTZ +zero)))))))
(assert
 (let ((?x14 (fp.sub roundNearestTiesToEven symbol0 symbol1)))
 (let ((?x28 (fp.sqrt roundNearestTiesToEven ?x14)))
 (and (distinct ?x28 ((_ to_fp 11 53) RTZ +zero)) true))))
(assert
 (fp.geq symbol0 ((_ to_fp 11 53) RTZ +zero)))
(assert
 (fp.leq (fp.abs (fp.sub roundNearestTiesToEven symbol0 symbol1)) ((_ to_fp 11 53) RTZ 1.9999999999999997779553950749686919152736663818359375 1023)))
(assert
 (not (fp.gt symbol1 ((_ to_fp 11 53) RTZ +zero))))
(check-sat)
```
In the translation, we used api `distinct()` for `neq` for floating-point constraints.

**Comment by wintersteiger** [comment]
Could you take a look at the model and propagate it through the constraints so that we see where things are going wrong? Thanks!

**Comment by delcypher** [comment]
@martin2011 Relating to what @wintersteiger just suggested. This is something I've had to do before in KLEE. If you're using the C API you might find the [validateModel()](https://github.com/klee/klee/blob/e3b88631ef58ad406ac069bd3a4ba16fb4aa07cc/lib/Solver/Z3Solver.cpp#L387) method in KLEE as a useful reference for how to do this.

**Comment by martin2011** [comment]
@delcypher , Thanks,  yet,  I do not use KLEE.

**Comment by delcypher** [comment]
@martin2011 
> @delcypher , Thanks, yet, I do not use KLEE.
That wasn't my point. I'm just telling you that if you want to check the model by propagating through the constraints like @wintersteiger suggests then you could do it by copying the approach used in KLEE.

**Comment by delcypher** [comment]
@martin2011 
> In the translation, we used api distinct() for neq for floating-point constraints.

This assumption might be wrong but it depends on semantics you expect from `neq`. If you expect to be the negation of `(fp.eq someVar someVar2)` then that is wrong and instead you should use `(not (fp.eq someVar someVar2))`.

This is because `(= someVar someVar2)` and `(fp.eq someVar someVar2)` are **not equivalent**. They have different semantics in the case that one or more of the operands is a NaN.

**Comment by martin2011** [comment]
@delcypher , Thanks.  Does that mean api  `dinstint()` can not apply in floating-point constraint for `x !=y` . I absolutely know `(= a b)` is not same as `(fp.eq a b)` and it does not exist an api for  floating-point `neq`. Thus, I try to use api `distinct()`.

**Comment by martin2011** [comment]
@wintersteiger A model was printed as followed:
```
(model 
  (define-fun symbol1 () (_ FloatingPoint 11 53)
    (fp #b0 #b00001111111 #x27e000fffffb8))
  (define-fun symbol0 () (_ FloatingPoint 11 53)
    (fp #b0 #b00010000011 #x013e000fffffd))
)
```
by running Z3 online: `http://rise4fun.com/z3/tutorial`. Can this model help?  It quitely hard for me  to reverify the constraints with bit-vector values. If necessary, I try to print a model from my project.

**Comment by LeventErkok** [comment]
@wintersteiger Perhaps tangential, but I'm curious what the following expression means in Z3:

```
((_ to_fp 11 53) RTZ 1.9999999999999997779553950749686919152736663818359375 1023))
```

(This appears in the OP's benchmark.)

It appears z3 interprets this as ` (fp #b0 #b11111111110 #xfffffffffffff)`, but if my reading of the SMTLib FP theory is correct, this should not be a valid expression: http://smtlib.cs.uiowa.edu/theories-FloatingPoint.shtml

Incidentally both MathSAT and CVC4 reject that expression as syntactically incorrect.

**Comment by delcypher** [comment]
> @delcypher , Thanks. Does that mean api dinstint() can not apply in floating-point constraint for x !=y . I absolutely know (= a b) is not same as (fp.eq a b) and it does not exist an api for floating-point neq. Thus, I try to use api distinct().

What semantics are you assigning to `x != y`? For example in the C programming language `NaN != NaN` is true where as `(distinct NaN NaN)` (not correct syntax but you get the idea)  is false.

**Comment by wintersteiger** [comment]
@martin2011 : You can add `(set-option :pp.fp_real_literals true)` to switch to real-numbered output of the model if that's easier for you to read.

@LeventErkok : Yes, that's not in any standard, but I use it for debugging as well sometimes. It means `1.9... * 2^1023` rounded according to RTZ. Quite possible that is has bugs though. 

`distinct` is the same as `(not (= ...` here; for instance:

```
(declare-fun x () (_ FloatingPoint 11 53))
(declare-fun y () (_ FloatingPoint 11 53))
(assert (distinct x y))
(apply simplify)
```
says
```
(goals
(goal
  (not (= x y))
  :precision precise :depth 1)
)
```

there is no `fp.neq`, but it's easy enough to define it locally:

```
(define-sort FPN () (_ FloatingPoint 11 53))
(define-fun fp.neq ((a FPN) (b FPN)) Bool (not (fp.eq a b)))

(assert (fp.neq x y))
...
```

**Comment by martin2011** [comment]
@wintersteiger . A model with real number is printed as followed:
```
(model 
  (define-fun symbol1 () (_ FloatingPoint 11 53)
    ((_ to_fp 11 53) RTZ 1.1557617783546287881790703977458178997039794921875 -896))
  (define-fun symbol0 () (_ FloatingPoint 11 53)
    ((_ to_fp 11 53) RTZ 1.0048522986471646323280992874060757458209991455078125 -892))
)
```
The result is interesting. By roungding-to-nearst, `sqrt(x-y)` seems to `underflow` and be rounding to be `zero`.  But, it does not satisfy the first sub condition of fifth constraint except the sub condition is checked from `x != y` and then `sqrt(x-y) !=0`.

**Comment by LeventErkok** [comment]
@martin2011 I think what's happening is that `symbol1 < symbol2`, and thus rounding takes it to `-0.0` (note the sign!), and then `sqrt` produces `NaN`. And since you're checking inequality with `distinct` (instead of `not fp.eq`), you're the correct but unexpected answer.

**Comment by martin2011** [comment]
@LeventErkok . There is no `symboli2`. The model has only `symbol1` and `symbol0`,  where`symbol0 > symbol1` and the constraint is `sqrt(symbol0 - symbol1 )!=0`.

**Comment by LeventErkok** [comment]
@martin2011 My bad. It's really hard to follow your encoding, perhaps you can try to manually "simplify" it first? That might help spot the issue better; regardless the problem might be in your encoding or in Z3.

**Comment by wintersteiger** [comment]
@martin2011 Are there any questions unanswered? Did you arrive at an encoding of your problem that behaves as expected, or can we close this issue?

**Comment by NikolajBjorner** [comment]
No answer yet. I suggest to close.

**Comment by wintersteiger** [comment]
@NikolajBjorner Good call. 

@martin2011 If something still needs resolving, please submit a new issue or ask us to reopen this one.
